### PR TITLE
feat(dashboards): add new endpoint for favourite dashboards

### PR DESCRIPTION
**Pull request by harshithadurai**
1. Add new endpoint for favouriting dashboards (supports `PUT` req)
2. Return `isFavourite` field in `DashboardDetails` endpoint's `GET` request

#78023

### CODE REVIEW

File: src/sentry/api/serializers/models/dashboard.py [hunk 1]
```diff
@@ -252,6 +254,7 @@ def serialize(self, obj, attrs, user, **kwargs) -> DashboardListResponse:
             "widgetDisplay": attrs.get("widget_display", []),
             "widgetPreview": attrs.get("widget_preview", []),
             "permissions": serialize(obj.permissions) if hasattr(obj, "permissions") else None,
+            "isFavourited": user.id in obj.favourited_by,
```
Comment (harshithadurai) [inline src/sentry/api/serializers/models/dashboard.py:257]: note to self: what happens when `favourited_by` doesn't exist in dashboard obj

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 2]
```diff
@@ -197,3 +197,48 @@ def post(self, request: Request, organization, dashboard) -> Response:
...
+
+        is_favourited = request.data.get("isFavourited", True)
+        current_favourites = set(dashboard.favourited_by)
+
+        if is_favourited and request.user.id not in current_favourites:
+            current_favourites.add(request.user.id)  # Add user to the set
+        elif not is_favourited and request.user.id in current_favourites:
+            current_favourites.remove(request.user.id)
+        else:
+            Response(status=204)
```
Comment (harshithadurai) [inline src/sentry/api/endpoints/organization_dashboard_details.py:239]: conditions:
not fav, not in list -> do nothing
not fav, in lits -> remove frm list
in fav, in list -> do nothing
in fav, not in list -> add to list

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 3]
```diff
@@ -197,3 +197,50 @@ def post(self, request: Request, organization, dashboard) -> Response:
...
+    """
+    Endpoint for managing the favorite status of dashboards for users
+    """
+
+    publish_status = {
+        "PUT": ApiPublishStatus.PRIVATE,
+        "GET": ApiPublishStatus.PRIVATE,
+    }
+
+    def get(self, request: Request, organization, dashboard) -> Response:
```
[the remark this answers is not in the document]
Comment (harshithadurai) [reply src/sentry/api/endpoints/organization_dashboard_details.py:213]: My idea was to use this when displaying the favourite status for an individual dashboard endpoint which will be used for the start here:
Screenshot 2024-12-02 at 1 53 29 PM

And the put request is used when clicking on the star

File: src/sentry/api/endpoints/organization_dashboard_details.py [hunk 4]
```diff
@@ -197,3 +197,50 @@ def post(self, request: Request, organization, dashboard) -> Response:
...
+    def put(self, request: Request, organization, dashboard) -> Response:
+        """
+        Toggle favorite status for current user by adding or removing
+        current user from dashboard favorites
+        """
+        if not features.has(EDIT_FEATURE, organization, actor=request.user):
+            return Response(status=404)
+
+        if isinstance(dashboard, dict):
+            return Response(status=204)
```
[the remark this answers is not in the document]
Comment (harshithadurai) [reply src/sentry/api/endpoints/organization_dashboard_details.py:232]: yup, it was for the pre-built dashboards. Would it make sense to keep this check in the GET and PUT methods, and not render the favourite option in the frontend? So, favouriting pre-built dashboards would just do nothing

File: src/sentry/api/endpoints/organization_dashboard_details.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (harshithadurai) [reply src/sentry/api/endpoints/organization_dashboard_details.py:213]: Instead, should I just add it to the `GET` of the `dashboard details` endpoint?

[the remark this answers is not in the document]
Comment (narsaynorath) [reply src/sentry/api/endpoints/organization_dashboard_details.py:213]: Yeah, I think we should just include it in the `GET` for dashboard details so we can cut down on the number of requests (we're already querying the dashboard with that request anyways). It'll just be ignored in the endpoint when there are updates

File: src/sentry/api/endpoints/organization_dashboard_details.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (narsaynorath) [reply src/sentry/api/endpoints/organization_dashboard_details.py:232]: We should check with Ale how we expect to handle the prebuilt dashboard, but I'm okay with leaving it in until we figure out what to do with it. If we don't want to let people favourite prebuilt dashboards (we only have one) then I think we should just 400 the request or something

**Review by narsaynorath — APPROVED** [review]
Just a comment about the feature flag, otherwise looks good!

**Pull request merged**